src/util/circular_buffer.c

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright Redis Ltd. 2016 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+
+#include "redismodule.h"
+#include "circular_buffer.h"
+#include "src/rmalloc.h"
+#include "src/util/likely.h"
+
+#include <stdatomic.h>
+
+// Circular buffer structure.
+// The buffer is of fixed size.
+// Items are removed by order of insertion, similar to a queue.
+struct _CircularBuffer {
+  char *read;                   // read data from here
+  _Atomic uint64_t write;       // write offset into data
+  size_t item_size;             // item size in bytes
+  _Atomic uint64_t item_count;  // current number of items in buffer
+  uint64_t item_cap;            // max number of items held by buffer
+  char *end_marker;             // marks the end of the buffer
+  char data[];                  // data
+};
+
+// Creates a new circular buffer, with `cap` items of size `item_size`
+CircularBuffer CircularBuffer_New(size_t item_size, uint cap) {
+  CircularBuffer cb = rm_calloc(1, sizeof(_CircularBuffer) + item_size * cap);
+
+  cb->read       = cb->data;                      // initial read position
+  cb->write      = ATOMIC_VAR_INIT(0);            // write offset into data
+  cb->item_cap   = cap;                           // buffer capacity
```

Comment:
Edge case: need to make sure it is not zero?
(may get to a wrong `write` offset in `CircularBuffer_Reserve`)

Reply:
Can you ellaborate?

Reply:
I see you handled zero cap or item_size.